### PR TITLE
fix & refactor: REST samples endpoints update + load_dotenv usage

### PR DESCRIPTION
**Pull request by GhiathAjam**
#### PR Classification
This pull request updates REST code samples (including endpoints) and improves robustness by using default values and clear error messages and supporting .env file for secrets.

#### PR Summary
- Environment variables are now loaded from the `.env` file using the `dotenv` package across all scripts.
- `.gitignore` has been updated to exclude the `.env` file, preventing accidental commits of sensitive data.
- The `endpoint` variable now has a default value depending on the service.
- `subscription_key` variable is fetched strictly from environment variables and throws a clear error message when it fails to load.
- Minor improvements:
    - `Entity Search` query changed to accommodate latest changes.
    - `Spell Check` now uses the better `POST` method instead of `GET`.

### CODE REVIEW

**Comment by GhiathAjam** [comment]
Please feel free to ignore, @hario90.

File: a
Comment (kerolloz) [inline a:1]: what is this file?

**Review by kerolloz — COMMENTED** [review]
Great job @GhiathAjam! ✨ 
My only take is that we might need to configure some formatting for the python source code in order to avoid changes like single and double quotes for string literals and spacing on variable assignments, like 

```python
API_KEY='BING_SEARCH_V7_AUTO_SUGGEST_SUBSCRIPTION_KEY'
# versus
API_KEY = 'BING_SEARCH_V7_AUTO_SUGGEST_SUBSCRIPTION_KEY'
```

**Comment by GhiathAjam** [comment]
> Great job @GhiathAjam! ✨ My only take is that we might need to configure some formatting for the python source code in order to avoid changes like single and double quotes for string literals and spacing on variable assignments, like
> 
> ```python
> API_KEY='BING_SEARCH_V7_AUTO_SUGGEST_SUBSCRIPTION_KEY'
> # versus
> API_KEY = 'BING_SEARCH_V7_AUTO_SUGGEST_SUBSCRIPTION_KEY'
> ```

Thanks Kerollos, that was actually discussed along with other changes and decided to be on a later PR.

**Review by kerolloz — APPROVED** [review]

**Comment by GhiathAjam** [comment]
[celebrate]     Gheiath Ajam (International Bu... reacted to your message:
________________________________

**Review by nader-m-elhadedy — COMMENTED** [review]
Awesome changes. Thanks Gheiath 👏 - Kindly find some comments below.


**Comments Summary**:
- missing [python-dotenv](https://pypi.org/project/python-dotenv/) package in requirements file
- other comments are suggestions

File: samples/rest/BingWebSearchV7.py [hunk 1]
```diff
@@ -7,33 +7,41 @@
...
+from dotenv import load_dotenv
+
+# Load the environment variables from .env file
+load_dotenv()
 
 '''
 This sample makes a call to the Bing Web Search API with a query and returns relevant web search.
 Documentation: https://docs.microsoft.com/en-us/bing/search-apis/bing-web-search/overview
 '''
+AUTH_HEADER=os.environ.get('BING_SEARCH_V7_AUTHORIZATION_HEADER', 'Ocp-Apim-Subscription-Key')
```
Comment (arwong) [inline samples/rest/BingWebSearchV7.py:19]: This is fine, but the subscription key header name hasn't changed since the Cognitive Services days and I don't foresee future changes there, so I think hard-coding the name is probably better.

File: samples/rest/BingWebSearchV7.py [hunk 2]
```diff
@@ -7,33 +7,41 @@
...
+API_KEY='BING_SEARCH_V7_WEB_SEARCH_SUBSCRIPTION_KEY'
```
Comment (arwong) [inline samples/rest/BingWebSearchV7.py:20]: Also OK but I think the variable names could be clearer.

For API_KEY it would be nice to indicate it's the name of an environment variable, and it's for subscription_key. Something like SUBSCRIPTION_KEY_ENV_VAR_NAME but that's pretty verbose, and elsewhere we're hard-coding environment variable names. -- maybe it'd be best to keep this hard-coded as well?

Same thing for AUTH_HEADER if we keep it an environment variable. Could we call it AUTH_HEADER_NAME?

File: samples/rest/BingWebSearchV7.py [hunk 3]
```diff
@@ -7,33 +7,41 @@
...
 
 # Add your Bing Search V7 subscription key and endpoint to your environment variables.
```
Comment (arwong) [inline samples/rest/BingWebSearchV7.py:22]: I suspect that the ability to set the endpoint URL via environment here is a holdover from Cognitive Services, where it was possible to customize the endpoint URL. That was not carried over into the current offering, so we should consider removing this option in the sample.

If we do so we should update the comments.

File: samples/rest/BingVideoSearchV7.py [hunk 4]
```diff
@@ -7,36 +7,45 @@
...
-    'Ocp-Apim-Subscription-Key': subscriptionKey
+    AUTH_HEADER: subscriptionKey
     }
-params = { "q": query }
+params = { 'q': query }
 
 # Call the API
 try:
-    response = requests.get(endpoint, headers=headers, params=params)
+    response = requests.get(endpoint + '/search', headers=headers, params=params)
```
Comment (arwong) [inline samples/rest/BingVideoSearchV7.py:41]: Related to my other comment about the customizability of the endpoint -- I think that given the features available today that either the endpoint is fully hard-coded, is fully-settable by environment variable and does not have a hard-coded /search component, or only really the hostname settable and that's the same environment variable among all the API samples.

**Review by arwong — CHANGES_REQUESTED** [review]
Glad to see this moving forward. I think this is close.

File: samples/rest/BingVideoSearchV7.py (hunk 4 above)
Comment (GhiathAjam) [reply samples/rest/BingVideoSearchV7.py:41]: The video search (and some others) actually has 3 endpoints, `/videos/search`, `/videos/trending`, and `/videos/details`. So I figured I add one sample for each endpoint that's why I hardcoded the last part. But after seeing your comments, I'm leaning towards either hardcoding the endpoints...

File: samples/rest/BingAutosuggestV7.py [hunk 5]
```diff
@@ -7,31 +7,39 @@
 import os
 import requests
 from pprint import pprint
+from dotenv import load_dotenv
+
+# Load the environment variables from .env file
+load_dotenv()
```
[the remark this answers is not in the document]
Comment (GhiathAjam) [reply samples/rest/BingAutosuggestV7.py:13]: Done: c2951726f2f2793197d097392e81c9d3c9a41b65
Added `python-dotenv` and `requests` to `requirements.txt`.
Note: `test_requirements.txt` will be updated in a later PR.

File: .gitignore [hunk 6]
```diff
@@ -1,3 +1,5 @@
+.env
+
```
[the remark this answers is not in the document]
Comment (GhiathAjam) [reply .gitignore:2]: Done: aaee62ec4862618af122513ffc88f05ec2669a79. Added `.env.example`.

This seemed the most popular when I searched. Correct me if I'm wrong.

**Comment by GhiathAjam** [comment]
@arwong @nader-m-elhadedy  Please re-review and resolve fixed issues. Thanks.

File: samples/rest/BingWebSearchV7.py (hunk 1 above)
Comment (GhiathAjam) [reply samples/rest/BingWebSearchV7.py]: Done: aaee62ec4862618af122513ffc88f05ec2669a79

File: samples/rest/BingWebSearchV7.py (hunk 2 above)
Comment (GhiathAjam) [reply samples/rest/BingWebSearchV7.py]: Done: aaee62ec4862618af122513ffc88f05ec2669a79

File: samples/rest/BingWebSearchV7.py (hunk 3 above)
Comment (GhiathAjam) [reply samples/rest/BingWebSearchV7.py]: Done: aaee62ec4862618af122513ffc88f05ec2669a79

File: samples/rest/BingAutosuggestV7.py [hunk 7]
```diff
@@ -7,31 +7,39 @@
...
+AUTH_HEADER=os.environ.get('BING_SEARCH_V7_AUTHORIZATION_HEADER', 'Ocp-Apim-Subscription-Key')
+API_KEY='BING_SEARCH_V7_AUTO_SUGGEST_SUBSCRIPTION_KEY'
 
 # Add your Bing Autosuggest subscription key and endpoint to your environment variables.
-subscription_key = os.environ['BING_AUTOSUGGEST_SUBSCRIPTION_KEY']
-endpoint = os.environ['BING_AUTOSUGGEST_ENDPOINT'] + '/bing/v7.0/Suggestions/'
+endpoint = os.environ.get('BING_SEARCH_V7_AUTO_SUGGEST_ENDPOINT', 'https://api.bing.microsoft.com/v7.0/suggestions')
+subscription_key = os.environ.get(API_KEY)
+if subscription_key is None:
+    raise(RuntimeError(f'Please define the {API_KEY} environment variable'))
```
[the remark this answers is not in the document]
Comment (GhiathAjam) [reply samples/rest/BingAutosuggestV7.py]: Will address in a later PR about refactoring REST samples for unit testing.

File: samples/rest/BingAutosuggestV7.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (nader-m-elhadedy) [reply samples/rest/BingAutosuggestV7.py:13]: Thanks for catching requests. Since it's a sub dependency (will be installed by default due to other dependency) but Explicit is better than implicit 👍

File: .gitignore (hunk 6 above)
[the remark this answers is not in the document]
Comment (nader-m-elhadedy) [reply .gitignore:2]: Yeah, it's good. Here is another example I found. [example](https://github.com/rudderlabs/rudder-sdk-python/blob/master/.env.sample)

Feel free to add equal sign only with empty value `ENV_VAR=` or `ENV_VAR=<env-var-value>` if this makes sense

**Review by arwong — APPROVED** [review]
LGTM

File: samples/rest/BingAutosuggestV7.py [hunk 8]
```diff
@@ -7,31 +7,39 @@
...
-headers = { 'Ocp-Apim-Subscription-Key': subscription_key }
+headers = { AUTH_HEADER_NAME: subscription_key }
 
 # Call the API
 try:
     response = requests.get(endpoint, headers=headers, params=params)
     response.raise_for_status()
 
-    print("\nHeaders:\n")
+    print('\nHeaders:\n')
```
Comment (Mohammed-Deifallah) [inline samples/rest/BingAutosuggestV7.py:39]: I suggest using and IDE linter to autoformat the code style on saving the file after edits.

File: samples/rest/BingSpellCheckV7.py [hunk 9]
```diff
@@ -6,23 +6,31 @@
...
-key = os.environ['BING_SPELL_CHECK_SUBSCRIPTION_KEY']
-endpoint = os.environ['BING_SPELL_CHECK_ENDPOINT'] + '/bing/v7.0/spellcheck'
+# Add your Bing Spell Check subscription key to your environment variables / .env file
+subscription_key = os.environ.get(SUBSCRIPTION_KEY_ENV_VAR_NAME)
+if subscription_key is None:
+    raise(RuntimeError(f'Please define the {SUBSCRIPTION_KEY_ENV_VAR_NAME} environment variable'))
 
 # Query you want spell-checked. 
-query = 'Hollo, wrld!'
+query = '''when i went two the houze i heared they'r'e voice and they're srcreams. I walk their and told: "helo fren"'''
```
Comment (Mohammed-Deifallah) [inline samples/rest/BingSpellCheckV7.py:28]: When i went two the house i heard their voices and their screams. I walked there and said: "hello fren"

File: samples/rest/BingSpellCheckV7.py [hunk 10]
```diff
@@ -33,13 +41,13 @@
...
+    print('\nHeaders:\n')
     print(response.headers)
 
-    print("\nJSON Response:\n")
+    print('\nJSON Response:\n')
     pprint(response.json())
 except Exception as ex:
-    raise ex
\ No newline at end of file
+    raise ex
```
Comment (Mohammed-Deifallah) [inline samples/rest/BingSpellCheckV7.py:53]: Check that link: https://stackoverflow.com/a/40280880/6941952
TL; DR You caught the exception to raise it as it is. You can do more by changing/customizing the error message, otherwise, the exception handling here is meant not to handle the exception.

**Review by Mohammed-Deifallah — APPROVED** [review]
I really appreciate your well-done work. This is impressive!
I have two general notes to take into consideration, but they are not mandatory for now:
1. Consider this python feature: https://docs.python.org/3/library/typing.html. I believe this will enhance your code readability and implementation experience.
2. I noticed that there are common pieces of code (or even the main flow) in `samples/rest/*.py`. Think about code redundancy.

File: samples/rest/BingAutosuggestV7.py (hunk 8 above)
Comment (GhiathAjam) [reply samples/rest/BingAutosuggestV7.py:39]: Formatting is reserved for another PR

File: samples/rest/BingCustomSearchV7.py [hunk 11]
```diff
@@ -9,23 +9,30 @@
...
-subscriptionKey = os.environ['BING_CUSTOM_SEARCH_SUBSCRIPTION_KEY']
-endpoint = os.environ['BING_CUSTOM_SEARCH_ENDPOINT']
-customConfigId = os.environ["BING_CUSTOM_CONFIG"]  # you can also use "1"
-searchTerm = "microsoft"
+searchTerm = 'Microsoft'
 # </importsAndVars>
 # <url>
-# Add your Bing Custom Search endpoint to your environment variables.
-url = endpoint + "/bingcustomsearch/v7.0/search?q=" + searchTerm + "&customconfig=" + customConfigId
+url = 'https://api.bing.microsoft.com/v7.0/custom/search?q=' + searchTerm + '&customconfig=' + customConfigId
```
[the remark this answers is not in the document]
Comment (GhiathAjam) [reply samples/rest/BingCustomSearchV7.py:33]: Do you think using an `f-string` would be appropriate here?

File: samples/rest/BingSpellCheckV7.py (hunk 9 above)
Comment (GhiathAjam) [reply samples/rest/BingSpellCheckV7.py:28]: hello fren, howiz itgoin

File: samples/rest/BingSpellCheckV7.py (hunk 10 above)
Comment (GhiathAjam) [reply samples/rest/BingSpellCheckV7.py:53]: This is left unchanged from earlier commits, will add "better exception handling" to the backlog, thanks

File: .env.example
[the remark this answers is not in the document]
Comment (GhiathAjam) [reply .env.example:1]: Is there a particular reason? or is this only a preference?

File: samples/rest/BingCustomSearchV7.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Mohammed-Deifallah) [reply samples/rest/BingCustomSearchV7.py:33]: As a coding practice, yes.

File: samples/rest/BingSpellCheckV7.py (hunk 9 above)
Comment (Mohammed-Deifallah) [reply samples/rest/BingSpellCheckV7.py:28]: Whatchulookinat :)

File: samples/rest/BingVisualSearchV7.py [hunk 12]
```diff
@@ -5,32 +5,42 @@
...
-# Add your Bing Search V7 subscriptionKey and endpoint to your environment variables.
-endpoint = os.environ['BING_SEARCH_V7_ENDPOINT'] + '/bing/v7.0/images/visualsearch'
-subscription_key = os.environ['BING_SEARCH_V7_SUBSCRIPTION_KEY']
+# Add your Bing Visual Search V7 subscription key to your environment variables / .env file
+subscription_key = os.environ.get(SUBSCRIPTION_KEY_ENV_VAR_NAME)
+if subscription_key is None:
+    raise(RuntimeError(f'Please define the {SUBSCRIPTION_KEY_ENV_VAR_NAME} environment variable'))
 
-image_path = 'MY-IMAGE' # for example: my_image.jpg
+image_path = './my_image.jpg'
```
[the remark this answers is not in the document]
Comment (Mohammed-Deifallah) [reply samples/rest/BingVisualSearchV7.py:26]: As long as it shows as a real path, you need to change it to an example path. In my opinion, it's better to use '<IMAGE-PATH>' instead

File: .env.example
[the remark this answers is not in the document]
Comment (Mohammed-Deifallah) [reply .env.example:1]: Sorry, check the original comment again.
It is better to minimize the template as much as you can to eliminate the confusion. Instead of writing '=' and the placeholder itself, you show the user that they need to only edit the placeholder instead of the whole line.

**Comment by Mohammed-Deifallah** [comment]
@GhiathAjam Yes, I guessed so, and you are 100% right.
I don't know if this is the case or not, but you can have some util functions, which are totally customized, and only import them in the samples provided. Again, you know better than me.
